### PR TITLE
New version of cmdlets for AzureRM.Billing and AzureRM.Consumption

### PR DESCRIPTION
**Pull request by Panda-Wang**
## Description

New version of cmdlets for AzureRM.Billing and AzureRM.Consumption
---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [X] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [X] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [X] Pull request includes test coverage for the included changes.
- [X] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [X] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [X] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [X] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [X] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [X] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Billing/Commands.Billing.Test/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -46,5 +46,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("0.11.0")]
-[assembly: AssemblyFileVersion("0.11.0")]
+[assembly: AssemblyVersion("0.12.0")]
+[assembly: AssemblyFileVersion("0.12.0")]
```
Comment (Panda-Wang) [inline src/ResourceManager/Billing/Commands.Billing.Test/Properties/AssemblyInfo.cs]: reverted

File: src/ResourceManager/Billing/Commands.Billing/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -46,5 +46,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("0.11.0")]
-[assembly: AssemblyFileVersion("0.11.0")]
+[assembly: AssemblyVersion("0.12.0")]
+[assembly: AssemblyFileVersion("0.12.0")]
```
Comment (Panda-Wang) [inline src/ResourceManager/Billing/Commands.Billing/Properties/AssemblyInfo.cs]: reverted

File: src/ResourceManager/Billing/AzureRM.Billing.psd1 [hunk 3]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '0.11.0'
+ModuleVersion = '0.12.0'
```
Comment (Panda-Wang) [inline src/ResourceManager/Billing/AzureRM.Billing.psd1]: reverted

File: src/ResourceManager/Billing/ChangeLog.md [hunk 4]
```diff
@@ -19,6 +19,13 @@
 -->
 ## Current Release
 
+## Version 0.12.0
```
Comment (Panda-Wang) [inline src/ResourceManager/Billing/ChangeLog.md]: removed

File: src/ResourceManager/Consumption/ChangeLog.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    * Overview of change #4
+        - Additional information about change #4
+
+    ## YYYY.MM.DD - Version X.Y.Z (Previous Release)
+    * Overview of change #1
+        - Additional information about change #1
+-->
+## Current Release
+
+## Version 0.1.0
```
Comment (Panda-Wang) [inline src/ResourceManager/Consumption/ChangeLog.md]: removed

File: src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/UsageDetails/GetAzureRmConsumptionUsageDetail.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Consumption.Cmdlets.UsageDetails
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmConsumptionUsageDetail", DefaultParameterSetName = Constants.ParameterSetNames.SubscriptionItemParameterSet), OutputType(typeof(List<PSUsageDetail>))]
+    public class GetAzureRmConsumptionUsageDetail : AzureConsumptionCmdletBase
+    {
+        const int MaxNumberToFetch = 1000;
```
Comment (Panda-Wang) [inline src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/UsageDetails/GetAzureRmConsumptionUsageDetail.cs]: this is still needed. This is the number of items to fetch in each request. as suggested by Mark, we keep pulling the next page until we have everything. but API has output size limit, thus we pull max 1000 result in each request to API call.

File: src/ResourceManager/Billing/Commands.Billing/Cmdlets/Invoices/GetAzureRmBillingInvoice.cs [hunk 7]
```diff
@@ -44,31 +44,46 @@ public class GetAzureRmBillingInvoice : AzureBillingCmdletBase
...
-                    throw new PSArgumentException(Resources.MaxCountExceedRangeError);
-                }
+                    string expand = this.GenerateDownloadUrl.IsPresent ? DownloadUrlExpand : null;
 
-                WriteObject(BillingManagementClient.Invoices.List(expand, null, null, MaxCount).Select(x => new PSInvoice(x)));
-                return;
-            }
+                    if (MaxCount.HasValue && (MaxCount.Value > 100 || MaxCount.Value < 1))
+                    {
+                        throw new PSArgumentException(Resources.MaxCountExceedRangeError);
```
Comment (Panda-Wang) [inline src/ResourceManager/Billing/Commands.Billing/Cmdlets/Invoices/GetAzureRmBillingInvoice.cs]: fixed

File: src/ResourceManager/Billing/Commands.Billing/Cmdlets/BillingPeriods/GetAzureRmBillingPeriod.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        [Parameter(Mandatory = false, HelpMessage = "Determine the maximum number of records to return.", ParameterSetName = Constants.ParameterSetNames.ListParameterSet)]
+        [ValidateNotNull]
+        public int? MaxCount { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            if (ParameterSetName.Equals(Constants.ParameterSetNames.ListParameterSet))
+            {
+                if (MaxCount.HasValue && (MaxCount.Value > 100 || MaxCount.Value < 1))
```
Comment (Panda-Wang) [inline src/ResourceManager/Billing/Commands.Billing/Cmdlets/BillingPeriods/GetAzureRmBillingPeriod.cs]: fixed

File: src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/UsageDetails/GetAzureRmConsumptionUsageDetail.cs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+        [Parameter(Mandatory = true, HelpMessage = "Name of a specific billing period to get the usage details that associate with.", ParameterSetName = Constants.ParameterSetNames.BillingPeriodItemParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public string BillingPeriodName { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Determine the maximum number of records to return.", ParameterSetName = Constants.ParameterSetNames.SubscriptionItemParameterSet)]
+        [Parameter(Mandatory = false, HelpMessage = "Determine the maximum number of records to return.", ParameterSetName = Constants.ParameterSetNames.InvoiceItemParameterSet)]
+        [Parameter(Mandatory = false, HelpMessage = "Determine the maximum number of records to return.", ParameterSetName = Constants.ParameterSetNames.BillingPeriodItemParameterSet)]
+        [ValidateNotNull]
+        public int? MaxCount { get; set; }
```
Comment (Panda-Wang) [inline src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/UsageDetails/GetAzureRmConsumptionUsageDetail.cs]: updated

**Comment by markcowl** [comment]
@Panda-Wang PR is failing because of the names of parameters ```IncludeMeterDetails``` and ```IncludeAdditonalProperties``` please add suppressions for these in the exceptions folder here:https://github.com/Azure/azure-powershell/blob/preview/tools/StaticAnalysis/Exceptions/SignatureIssues.csv (you can just copy them from the static analysis reports here: http://azuresdkci.cloudapp.net/job/powershell/8291/artifact/src/Package/SignatureIssues.csv)

**Comment by markcowl** [comment]
@azuresdkci retest this please

**Comment by markcowl** [comment]
@Panda-Wang You will need to pull the latest and rebuild your wxi file.  BTW - are you adding new dependency files?  If not, you don't need to update the wxi.

**Comment by markcowl** [comment]
@Panda-Wang It looks like you have added a bunch of changed files outside your own changes.  Can you please fetch the latest locally before updating your PR

**Comment by Panda-Wang** [comment]
Re-submitted from a clean fork: https://github.com/Azure/azure-powershell/pull/3880

**Pull request closed without merging**